### PR TITLE
[Golden Gate] BlackBox-based Endpoint System

### PR DESCRIPTION
**Pull request by davidbiancolin**
This PR removes the Chisel-type driven endpoint generation system in favor of a BlackBox-based one. Instead of providing an endpoint map, the user must bind endpoint BlackBoxes to all I/O in their system, to create a closed Chisel system that describes the target. This more cleanly separates it from the target from the host-level transformations MIDAS preforms. 

This will enable the following:
- Serialization of the target's FIRRTL before MIDAS consumes it, making it easier to deploy FireSim as a library
  - endpoint configuration can partially determined at target-elaboration time, and partially during platform mapping, creating better seperation of target and host configuration (TARGET_CONFIG vs PLATFORM_CONFIG)
- The ability to embed endpoints in the target's module hierarchy. 
  - the user does not need to use a wiring transform to bind an endpoint to an interface deep in the module hierarchy. 
  - example use case: can write and integrate a SW RoCC model by connecting an Endpoint BlackBox to the interface
- Make it easier to instantiate a variety of different endpoints of the same Module class with different  parameterizations
  - ex. Heterogeneous memory channels.

### CODE REVIEW

File: src/main/scala/midas/passes/fame/FAMEUtils.scala [hunk 1]
```diff
@@ -19,7 +19,6 @@ object RTRenamer {
     { rt =>
       val renameMatches = renames.get(rt).getOrElse(Seq(rt)).collect({ case rt: ReferenceTarget => rt })
       if (renameMatches.length == 0)
-        println(rt)
```
Comment (albert-magyar) [inline src/main/scala/midas/passes/fame/FAMEUtils.scala]: Is the `if` supposed to be guarding the assert now?

File: src/main/scala/midas/passes/fame/PromoteSubmodule.scala [hunk 2]
```diff
@@ -89,7 +89,7 @@ class PromoteSubmodule extends Transform {
     val order = reversedIGraph.linearize.filter(reversedIGraph.getEdges(_).size > 0).filter(promoted)
     val renames = RenameMap()
     for (childInstance <- order) {
-      val childModule = updatedModules(childInstance.module)
+      val childModule = iGraph.moduleMap(childInstance.module)
```
Comment (albert-magyar) [inline src/main/scala/midas/passes/fame/PromoteSubmodule.scala]: Why do you want the un-updated child module?

File: src/main/scala/midas/passes/EnsureNoTargetIO.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import firrtl.transforms.TopWiring.{TopWiringAnnotation, TopWiringTransform, TopWiringOutputFilesAnnotation}
+import firrtl.passes.wiring.{Wiring, WiringInfo}
+import Utils._
+
+import scala.collection.mutable
+import java.io.{File, FileWriter, StringWriter}
+
+// Ensures that there are no dangling IO on the target. All I/O coming off the DUT must be bound
+// to an Endpoint BlackBox
+private[passes] class EnsureNoTargetIO extends firrtl.Transform {
```
Comment (albert-magyar) [inline src/main/scala/midas/passes/EnsureNoTargetIO.scala]: 👍

File: src/main/scala/midas/passes/ExtractEndpoints.scala [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    val endpointInsts: Seq[Seq[WDefInstance]] = endpointModules.flatMap(e => iGraph.findInstancesInHierarchy(e).map(_.reverse))
+
+    // Generate instance annotations to drive promoteEndpoints()
+    val instAnnos = endpointInsts.collect({ case endpoint :: parent :: restOfPath =>
+      EndpointInstance(InstanceTarget(c.main, parent.module, Nil, endpoint.name, endpoint.module))
+    })
+    state.copy(annotations = state.annotations ++ instAnnos)
+  }
+
+  def getEndpointConnectivity(portInstMapping: mutable.ArrayBuffer[(String, String)],
```
Comment (albert-magyar) [inline src/main/scala/midas/passes/ExtractEndpoints.scala]: You can use `foreach` from `firrtl.traversals.Foreachers` for this

**Comment by davidbiancolin** [comment]
Closing to change base branch

**Pull request closed without merging**